ocaml/xapi/xapi_vm_migrate.ml

```diff
@@ -227,6 +228,18 @@ let rec migrate_with_retries ~__context queue_name max try_no dbg vm_uuid
           "xenops: not retrying migration: caught %s from %s in attempt %d of \
            %d."
           (Printexc.to_string e) !progress try_no max ;
+        ( try
+            Xapi_vm_lifecycle.force_state_reset ~__context ~self:vm
```

Comment:
Should we add a safety check here that the VM on the source is not running (i.e. paused, etc.)? If it is still running then there is no need to kill it, we can just let it keep running, the migration failed but we haven't made any permanent changes to the source VM yet.

Reply:
This is a dangerous operation, and I don't think it does what you want.

We should let xenopsd handle the cleanup, in the xenopsd PR. If that starts working, then xapi will update the VM state through the usual xenopsd event mechanism.

Reply:
There is also existing fallback cleanup code in the exception handlers in `pool_migrate` and `migrate_send`, using `Xapi_xenops.shutdown` to give xenopsd another kick to clean up if required.